pkg/agent/controller/trafficcontrol/controller.go

```diff
@@ -0,0 +1,1009 @@
+		// If the name is different from the cached name in the TrafficControl state, it could be caused by the return
+		// port update of the TrafficControl or the creation of the TrafficControl.
+		if returnPortName != tcState.returnPortName {
+			if tcState.returnPortName != "" {
+				// If the stale return port name cached in TrafficControl state is not empty, release the stale return port
+				// from the TrafficControl.
+				returnOFPort, portDeleted, err := c.releaseTrafficControlPort(returnPortName, tcName)
+				if err != nil {
+					return err
+				}
+				// If the stale return port is deleted, uninstall the return flow for the return port.
+				if portDeleted {
+					if err = c.ofClient.UninstallTrafficControlReturnPortFlow(returnOFPort); err != nil {
+						return err
+					}
+				}
+				tcUpdated = true
+			}
+			// Get or create the return port.
+			var returnOFPort uint32
+			if returnOFPort, exists, err = c.getOrCreateTrafficControlPort(tc.Spec.ReturnPort, returnPortName, tcName); err != nil {
+				return err
+			}
+			// If the return port is newly created, install a return flow for the return port.
+			if !exists {
+				if err = c.ofClient.InstallTrafficControlReturnPortFlow(returnOFPort); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	// Get name of the return port.
+	targetPortName = c.getPortName(&tc.Spec.TargetPort)
+	// If the name is different from the cached name in the TrafficControl state, it could be caused by the target port
+	// update of the TrafficControl or the creation of the TrafficControl.
+	if targetPortName != tcState.targetPortName {
+		if tcState.targetPortName != "" {
+			// If the stale return port name cached in TrafficControl state is not empty, release the stale target port
+			// from the TrafficControl.
+			_, _, err = c.releaseTrafficControlPort(tcState.targetPortName, tcName)
+			if err != nil {
+				return err
+			}
+			tcUpdated = true
+		}
+	}
+	// Get or create the target port.
+	targetOFPort, _, err := c.getOrCreateTrafficControlPort(&tc.Spec.TargetPort, targetPortName, tcName)
+	if err != nil {
+		return err
+	}
+
+	// Update the TrafficControl state.
+	tcState.targetPortName = targetPortName
+	tcState.returnPortName = returnPortName
+	if tcState.action != tc.Spec.Action {
+		tcState.action = tc.Spec.Action
+		tcUpdated = true
+	}
+	if tcState.direction != tc.Spec.Direction {
+		tcState.direction = tc.Spec.Direction
+		tcUpdated = true
+	}
+
+	// Get the list of Pods applying to the TrafficControl.
+	var podObjects []*v1.Pod
+	if podObjects, err = c.filterPods(&tc.Spec.AppliedTo); err != nil {
+		return err
+	}
+
+	// Reserve the set of Pods in TrafficControl state as stale, and reinitialize the set in TrafficControl state.
+	stalePods := tcState.pods
+	tcState.pods = sets.NewString()
+	// Reserve the set of OF ports in TrafficControl state as old, and reinitialize the set in TrafficControl state.
+	oldOfPorts := tcState.ofPorts
+	tcState.ofPorts = sets.NewInt32()
```

Comment:
ditto, we should update tcState after the ofPort's flow has been installed?

Reply:
Done